src/Processors/Executors/PipelineExecutor.cpp

```
@@ -303,64 +311,57 @@ void PipelineExecutor::initializeExecution(size_t num_threads)
-            {
-                executeSingleThread(thread_num);
-            }
-            catch (...)
+            pool.scheduleOrThrowOnError([this, thread_num, thread_group = CurrentThread::getGroup(), slot = std::move(slot)]
             {
-                /// In case of exception from executor itself, stop other threads.
-                finish();
-                tasks.getThreadContext(thread_num).setException(std::current_exception());
-            }
-        });
-    }
-}
+                /// ThreadStatus thread_status;
 
-void PipelineExecutor::joinThreads()
-{
-    for (size_t thread_num = 0; ; thread_num++)
-    {
-        std::unique_lock lock{threads_mutex};
-        if (thread_num >= threads.size())
-            break;
-        if (threads[thread_num].joinable())
+                SCOPE_EXIT_SAFE(
+                    if (thread_group)
+                        CurrentThread::detachFromGroupIfNotDetached();
+                );
+                setThreadName("QueryPipelineEx");
+
+                if (thread_group)
+                    CurrentThread::attachToGroup(thread_group);
+
+                try
+                {
+                    executeSingleThread(thread_num);
+                }
+                catch (...)
+                {
+                    /// In case of exception from executor itself, stop other threads.
+                    finish();
+                    tasks.getThreadContext(thread_num).setException(std::current_exception());
+                }
+            });
+        }
+        catch (...)
         {
-            auto & thread = threads[thread_num];
-            lock.unlock(); // to avoid deadlock if thread we are going to join starts spawning threads
-            thread.join();
+            /// In case of exception from executor itself, stop other threads.
+            finish();
+            tasks.getThreadContext(thread_num).setException(std::current_exception());
+
+            pool.wait();
+            throw;
```

And now we save the exception, throw it again, and also save it again 10 lines above.

I think this try/catch is not needed at all. For a single thread, we can throw an exception as is. For multiple threads, we handle exceptions above in lambda.

---

> I think this try/catch is not needed at all. For a single thread, we can throw an exception as is. For multiple threads, we handle exceptions above in lambda.

Agree, removed.